### ISSUE TITLE
Partially unpacked Steam Linux Runtime does not run and gives unclear log hints

### ISSUE THREAD

**Issue by julien-f** [ValveSoftware/steam-runtime#300]
# Compatibility Report

No games are working, but we can say Deep Rock Galactic (548430) or Risk of Rain 2 (632360)

## System Information
- GPU: NVIDIA Corporation GeForce GTX 970/PCIe/SSE2
- Driver/LLVM version: 4.6.0 NVIDIA 455.38
- Kernel version: 5.8.18-200.fc32.x86_64
- Link to full system information report as [Gist](https://gist.github.com/): https://gist.githubusercontent.com/julien-f/62dcd927ff10d425676703d73396423a/raw/b0c4610411cf552f335d7bd57137d5d05b80d593/systemInfo.txt
- Proton version: 5.13-2

I am using *Steam Beta Update*, I have not installed Proton manually nor customized it in anyway.

## I confirm:
- [x] that I haven't found an existing compatibility report for this game. (probably related to ValveSoftware/Proton#4345 
- [x] that I have checked whether there are updates for my system available.

Unfortunately, I did not find this file after adding the env variable.

I ran `steam` with this command and got this trace: https://gist.githubusercontent.com/julien-f/62dcd927ff10d425676703d73396423a/raw/b0c4610411cf552f335d7bd57137d5d05b80d593/steam.log

```
PRESSURE_VESSEL_VERBOSE=1 PROTON_LOG=1 steam 2>&1 | tee ~/proton-issue.log
```

## Symptoms 

1. I click on Play
2. Shaders are compiled
3. Nothing happens and Steam reports this game as closed

**Comment by kisak-valve** [comment]
Hello @julien-f, let's treat this as a Pressure Vessel issue until there's a stronger indication that the issue is elsewhere.

**Comment by julien-f** [comment]
I have no idea no what it is 😅 
Is there anything I should do?

**Comment by TTimo** [comment]
```
pressure-vessel-wrap[15639]: Replacing self with bwrap...
bwrap: Can't mkdir /usr/lib64/gconv: Read-only file system
ln: failed to create symbolic link '/tmp/SteamPVSockets.ZYIIly/SteamLinuxRuntime.f6e06be2d6a99c5a/socket' -> '': No such file or directory
```
```
CClientJobGatherRuntimeInformation: container read output failed
```
We've seen this when the soldier SLR depot is corrupted in some way.

Try this: in your Steam library, find `Steam Linux Runtime - Soldier`. In properties, do Manage > Uninstall, and then install it again. Restart Steam afterwards. Make sure you no longer have this `CClientJobGatherRuntimeInformation: container read output failed` message at the end of your Help > System Information output.

**Comment by julien-f** [comment]
I have uninstalled and reinstalled Soldier and restarted Steam, but I still have this error at the end of the system info.

**Comment by TTimo** [comment]
Please add `G_MESSAGE_DEBUG=all` to your environment and capture the logs again. We are looking for a line from pressure-vessel-launcher that says `Chosen socket is`

You can also add a line of verbose log at the end of the `_start-container-in-background` script (in the `SteamLinuxRuntime_soldier` depot) like so (expectation is that source will be empty for you):

```
log "TTimo - source: $socket link name: $rendezvous/socket"
ln -fns "$socket" "$rendezvous/socket"
verbose "OK, ready to exit"
exit 0
```

Something is going wrong with the creation (or passing) of the rendezvous socket.

**Comment by julien-f** [comment]
> Please add `G_MESSAGE_DEBUG=all` to your environment and capture the logs again. We are looking for a line from pressure-vessel-launcher that says `Chosen socket is`

I've run the following command but I'm unable to find the line: https://gist.githubusercontent.com/julien-f/62dcd927ff10d425676703d73396423a/raw/4fe44ab61b67aa85b909fcf5b37649ebe59bb21b/steam2.log

```
env G_MESSAGE_DEBUG=all PRESSURE_VESSEL_VERBOSE=1 PROTON_LOG=1 steam 2>&1 | tee ~/proton-issue.log
```


> You can also add a line of verbose log at the end of the `_start-container-in-background` script (in the `SteamLinuxRuntime_soldier` depot) like so (expectation is that source will be empty for you):

I don't know how to do this, I don't know what is the `SteamLinuxRuntime_soldier` depot, I've tried to find the file `_start-container-in-background` somewhere in `~/.steam` with no success.

FYI, I did not install anything by hand so far.

**Comment by TTimo** [comment]
Thanks - looks like bubblewrap on your host is not behaving as we'd expect (`bwrap`):

```
  pressure-vessel-wrap[175839]: Checking for x86_64-linux-gnu libraries...
  pressure-vessel-wrap[175839]: run: '/home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/pressure-vessel/bin/x86_64-linux-gnu-capsule-capture-libs' '--print-ld.so'
  pressure-vessel-wrap[175839]: -> /lib64/ld-linux-x86-64.so.2
  pressure-vessel-wrap[175839]: run: '/usr/bin/bwrap' '--ro-bind' '/home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/soldier_0.20201022.1/files' '/usr' '--symlink' 'usr/lib' '/lib' '--symlink' 'usr/.ref' '/.ref' '--symlink' 'usr/bin' '/bin' '--ro-bind' '/usr' '/run/host/usr' '--symlink' 'usr/bin' '/run/host/bin' '--symlink' 'usr/lib' '/run/host/lib' '--symlink' 'usr/sbin' '/run/host/sbin' '--symlink' 'usr/lib64' '/run/host/lib64' '--ro-bind' '/etc/alternatives' '/run/host/etc/alternatives' '--ro-bind' '/etc/ld.so.cache' '/run/host/etc/ld.so.cache' '--ro-bind' '/etc/ld.so.conf' '/run/host/etc/ld.so.conf' '--ro-bind' '/etc/ld.so.conf.d' '/run/host/etc/ld.so.conf.d' 'env' 'PATH=/usr/bin:/bin' 'readlink' '-e' '/lib64/ld-linux-x86-64.so.2'
  bwrap: execvp env: No such file or directory
  pressure-vessel-wrap[175839]: Container does not have /lib64/ld-linux-x86-64.so.2 so it cannot run x86_64-linux-gnu binaries
  pressure-vessel-wrap[175839]: Checking for i386-linux-gnu libraries...
  pressure-vessel-wrap[175839]: run: '/home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/pressure-vessel/bin/i386-linux-gnu-capsule-capture-libs' '--print-ld.so'
  pressure-vessel-wrap[175839]: -> /lib/ld-linux.so.2
  pressure-vessel-wrap[175839]: run: '/usr/bin/bwrap' '--ro-bind' '/home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/soldier_0.20201022.1/files' '/usr' '--symlink' 'usr/lib' '/lib' '--symlink' 'usr/.ref' '/.ref' '--symlink' 'usr/bin' '/bin' '--ro-bind' '/usr' '/run/host/usr' '--symlink' 'usr/bin' '/run/host/bin' '--symlink' 'usr/lib' '/run/host/lib' '--symlink' 'usr/sbin' '/run/host/sbin' '--symlink' 'usr/lib64' '/run/host/lib64' '--ro-bind' '/etc/alternatives' '/run/host/etc/alternatives' '--ro-bind' '/etc/ld.so.cache' '/run/host/etc/ld.so.cache' '--ro-bind' '/etc/ld.so.conf' '/run/host/etc/ld.so.conf' '--ro-bind' '/etc/ld.so.conf.d' '/run/host/etc/ld.so.conf.d' 'env' 'PATH=/usr/bin:/bin' 'readlink' '-e' '/lib/ld-linux.so.2'
  bwrap: execvp env: No such file or directory
  pressure-vessel-wrap[175839]: Container does not have /lib/ld-linux.so.2 so it cannot run i386-linux-gnu binaries
  pressure-vessel-wrap[175839]: None of the supported CPU architectures are common to the host system and the container (tried: x86_64-linux-gnu, i386-linux-gnu)
  pressure-vessel-wrap[175839]: Exiting with status 1
```

Can you temporarily disable the host bwrap (`mv /usr/bin/bwrap /usr/bin/bwrap.off`), or uninstall the `bubblewrap` package and try again? Looks like fedora is setup for setuid bwrap so I don't think that will fix but we might learn something.

**Comment by julien-f** [comment]
No success after renaming `bwrap` (could not uninstall as Flatpak depends on it): https://gist.githubusercontent.com/julien-f/62dcd927ff10d425676703d73396423a/raw/9e04081b41d78e1f3889ddc459a714e840106d65/steam3.log

**Comment by TTimo** [comment]
You're not running steam flatpak are you - your logs didn't suggest that would be the case.

**Comment by julien-f** [comment]
> You're not running steam flatpak are you - your logs didn't suggest that would be the case.

Nope (and I've seen the corresponding issue) but you're right to ask :slightly_smiling_face:

**Comment by smcv** [comment]
> bwrap: execvp env: No such file or directory

This is likely to be part of the root cause. I don't know what would be causing the container to not have an `env` in `PATH`, but I'm not surprised it breaks things.

@julien-f, what does the `PATH` on the host system look like?

> bwrap: Can't mkdir /usr/lib64/gconv: Read-only file system

I've seen this reported before, but I'm not sure what causes it. If this happens, it'll also break the container.

We pre-create `usr/lib64/gconv` in the container runtime (it's `soldier_0.x.y/files/lib64/gconv/` on disk), so it shouldn't be possible for this to happen...

> Looks like fedora is setup for setuid bwrap

I don't think so? I think modern Fedora's `/usr/bin/bwrap` is non-setuid, like Ubuntu.

The only major distributions I'm aware of that normally need a setuid bwrap are Debian, and Arch Linux with a non-default kernel.

**Comment by smcv** [comment]
We can try using `/usr/bin/env` explicitly, instead of `env`, when we run the `env` in the container. 99% of host systems should have `/usr/bin/env`, and our containers definitely have that path.

**Comment by julien-f** [comment]
> @julien-f, what does the `PATH` on the host system look like?

```
> echo "$PATH"
/home/julien/.local/bin:/home/julien/bin:/home/julien/.cargo/bin:/home/julien/.local/bin:/home/julien/bin:/home/julien/.cargo/bin:/home/julien/.local/bin:/home/julien/bin:/usr/share/Modules/bin:./node_modules/.bin:/home/julien/dev/julien-f/config/scripts:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/julien/n/bin
> which env
/usr/bin/env
```

**Comment by julien-f** [comment]
> I don't think so? I think modern Fedora's `/usr/bin/bwrap` is non-setuid, like Ubuntu.

I'm using Fedora 32 and `bwrap` is not setuid.

```
> ls -l /usr/bin/bwrap
-rwxr-xr-x. 1 root root 74K Mar 30  2020 /usr/bin/bwrap*
```

**Comment by smcv** [comment]
Yeah, I don't get it...

Please try renaming `/home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier/soldier_0.20201022.1` to something like `soldier_bug300`, and create a file `soldier_bug300/keep` so that it won't be deleted. Then try to launch the container:

```
cd /home/julien/.local/share/Steam/steamapps/common/SteamLinuxRuntime_soldier
PRESSURE_VESSEL_VERBOSE=1 ./run xterm 2>&1 | tee ~/bug300-new.log
```

I'm hoping that will work. If it works, an xterm will pop up with a shell inside the container.

If that is successful, then please compare `soldier_bug300` with the new `soldier_0.20201022.1` directory:

```
diff -r --brief soldier_bug300 soldier_0.20201022.1
```

**Comment by TTimo** [comment]
Oh that's right, I asked to uninstall/reinstall soldier SLR previously, but I forgot the Steam client just leaves non depot files in place. So corrupted tarball expansions would persist.

**Comment by julien-f** [comment]
`xterm` opens and I get this diff: https://gist.githubusercontent.com/julien-f/62dcd927ff10d425676703d73396423a/raw/74e426edd1156a0729a6acec4c479af0879c0a8d/soldier_bug300.diff

**Comment by julien-f** [comment]
Just doing this manipulation fixed it for me!

**Comment by smcv** [comment]
OK. So the problem here seems to be: on a previous attempt to launch the container, the archive was only partially unpacked, but then it was renamed into place anyway (? I'm not sure how this can happen), and then launching the partially-unpacked version was broken from then on.

**Comment by smcv** [comment]
We unpack the archive into a temporary directory, and only rename that temporary directory to its final name after it has been unpacked successfully, so I don't really understand how that can have failed. We also take a lock on the `SteamLinuxRuntime_soldier` directory while we are unpacking the new runtime, so even if you somehow run two games concurrently, they shouldn't be able to fight over it: the first one should finish before the second one can start.

In an ideal world, we'd deploy the runtime in the obvious way as part of the depot, but the steampipe system is designed to deal with a few relatively large files, and the runtime is a lot of relatively small files, so that didn't work well when we tried it. Instead, we ship it as a big tar archive, and unpack it as-needed (Proton uses the same trick).

The next best thing from a robustness point of view would be to unpack a new copy of the whole runtime for each run of a game, but that would be really slow;  so we cache a copy, and only re-unpack it when there has been a version update.

We can't ship the runtime as squashfs or something and mount it as a filesystem, because that would need root (or at best, FUSE, a performance penalty, and another thing that can go wrong on unusually-configured host systems).

At some point I'll probably move the cache of unpacked runtime files into a `var/` subdirectory (or maybe into compatdata, or something), which would mean Steam could potentially delete that entire directory as part of a cleanup step when reinstalling or verifying the depot. Maybe that would help?

**Comment by suamor** [comment]
It seems I have the same issue because I also get a broken soldier. I'm on Ubuntu 20.04 LTS. I use official steam version and previously made a complete new installation. I definitively see diff between the original installed and the "manually" extracted soldier. However (this is the reason why I write here), after the "manipulation", i.e. the manual extraction of the archive I'm still not able to start soldier but get a clean crash when trying to get system information from steam and scrolling to the end (the error message remains the same). So this issue looks to be more complex than just a broken installation in my case.

**Comment by suamor** [comment]
Update from my side: As I still had the crash with proton GE but fixed a different issue namely a problem with my "/home" link which proton for some reason does not like (and replaced my $HOME with my physical home folder), I found that the culprit for the crash is bbswitch. It simply fails to work with nvidia driver. because bumblebee is unmaintained since years, I completely purged it. Instead I use now "glxhead" for switching. I still have to do some tests, but the crash issue is definitively a problem with graphics driver, not something with the archive.

**Comment by smcv** [comment]
> a problem with my "/home" link which proton for some reason does not like (and replaced my $HOME with my physical home folder)

Please report this as a separate issue, unless you consider it to be local misconfiguration that we can safely ignore.

**Comment by smcv** [comment]
Versions >= 0.20210503.0 (currently only in the `client_beta` branch) have switched to a different way of distributing the runtime which should mean this failure mode can't happen.

**Comment by kisak-valve** [comment]
Tracking note: Steam Linux Runtime (scout) mainline release has been bumped to 0.20210518.3 and Steam Linux Runtime - Soldier is at 0.20210415.0, so we're waiting on the latter to get updated past 0.20210503.0 to close this issue report.

**Comment by kisak-valve** [comment]
Steam Linux Runtime - Soldier mainline release has been bumped to 0.20210629.0. We're done here.
